rfcs/text/0014_notification_service.md

```diff
@@ -0,0 +1,401 @@
+- *RemoteRepository* - a repository for the Cloud version of Kibana. It stores data within the Unified Notification System deployed in the Cloud.
+In this case, a user can see in Kibana UI the external notifications created outside of Kibana.
+
+*Repository* interface:
+- `create(NotificationEvent)`- creates a notification in the Notification system. 
+- `get({ search_after?: number; size?: number; filter?: NotificationFilter })` - retrieves last notifications for the current user.
+Might be paginated using `search_after` filed.
+- `markAsRead({ notification_id: string[] })` - mark notifications with given ids as read.
+- `markAsUnread({ notification_id: string[] })` - mark notifications with given ids as unread.
+- `markAllAsRead()` - mark all the notifications as read.
+- `markAsPinned({ notification_id: string[] })` - mark notifications with given ids as pinned.
+- `markAsUnpinned({ notification_id: string[] })` - mark notifications with given ids as unpinned.
+- `configure()` - updates user-specific notification settings.
+
+*Repository* interace if not exposed outside of Notification system.
+Sources interact with Notification sustem via plugin contract:
+```typescript
+interface Notifications {
+  create(events: NotificationEvent): void;
+  createForAllUsers(events: Omit<NotificationEvent, 'recipient_id'>): void;
+}
+```
+
+#### LocalRepository
+![image](../images/0014/local_repository.png)
+
+##### Notification model
+Notification model is responsible for:
+- Applying policies to an incoming notification. NS must check whether a recipient unsubscribed from a source before processing to delivery.
+- Enhancing an incoming notification with necessary data: 
+    - `expire_at` if not specified. Kibana allows administrators to set how long notifications will be stored. Falls back to 30 - 90 days be default. 
+- Writing new incoming messages to the Notification storage.
+
+Processing a large number of incoming notifications might slow down Kibana and create a disproportionately high load on some Kibana instances.
+The use of a queue allows us to make the handling of incoming notifications asynchronous.
+Async mechanism lets Kibana to cope with spikes in the number of messages created. The queue implementation is based on the Elasticsearch server. 
+
+The logic is run as a background task. The usage of the single background task:
+- reduces the possibility of read- and write-conflicts
+- allows re-using task_manager built-in features: scheduling, retry, monitoring
+
+If Kibana processes notifications in asynchronous batches, we will need a prioritization mechanism to be sure events
+with a higher severity are delivered first without a significant delay. The prioritization mechanism is based on
+priority property value: the notification with the higher value is processed earlier to provide an experience as close to real-time as possible.
+
+##### Notification storage
```

Comment:
How are global notifications going to be persisted? Are they going to be per-user "marked as read", etc.?

Reply:
> How are global notifications going to be persisted? Are they going to be per-user "marked as read", etc.?

It's described in [workflow](https://docs.google.com/document/d/18s-BTHog_oPaQKf871gzuhxkah4pK1GYJM9DCLceaLo/edit#heading=h.3mvk0hq96lm7).

We will store global notifications without user-specific status. Every user has their own NotificationStatus object per a global notification.
```typescript
interface  NotificationState {
  notification_id: string;
  recipient_id: string;
  is_read: boolean;
  is_pinned: boolean;
}
````
User-specific `NotificationState` SO object is created only when a user changes `NotificationState`.

Reply:
Cool, thanks @restrry